<a href="https://colab.research.google.com/github/chirag-trivedi/data-science/blob/main/MovieLens.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

Read the dataset

In [2]:
movie = pd.read_csv("movie.csv")
user = pd.read_csv("user.csv")
rating = pd.read_csv("ratings.csv")

In [3]:
user.head()

user id  age gender  occupation zip code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

We observe that the user data set has 5 columns user id , age , gender , occupation and zip code.

In [4]:
user.shape

(943, 5)

We can see that there are 943 rows and 5 cols

In [5]:
user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user id     943 non-null    int64 
 1   age         943 non-null    int64 
 2   gender      943 non-null    object
 3   occupation  943 non-null    object
 4   zip code    943 non-null    object
dtypes: int64(2), object(3)
memory usage: 37.0+ KB


We can see that gender, occupation and zip code are of the object type while user id and age are ints.The zip code data type seems to be wrong

In [6]:
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

The average age is 34 years with a median of 31 years.THe range for age is between 7 and 73 years.stats for user id are meaning less.

**Movie Data Set**

In [7]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

Here is the movie data set . We can see that we have an id , title and release date for the movie and the genre .A movie can belong to multiple genres.

In [8]:
movie.shape

(1680, 21)

There are 1680 rows and 21 cols

In [9]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1680 entries, 0 to 1679
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   movie id      1680 non-null   int64 
 1   movie title   1680 non-null   object
 2   release date  1680 non-null   object
 3   Action        1680 non-null   int64 
 4   Adventure     1680 non-null   int64 
 5   Animation     1680 non-null   int64 
 6   Childrens     1680 non-null   int64 
 7   Comedy        1680 non-null   int64 
 8   Crime         1680 non-null   int64 
 9   Documentary   1680 non-null   int64 
 10  Drama         1680 non-null   int64 
 11  Fantasy       1680 non-null   int64 
 12  Film-Noir     1680 non-null   int64 
 13  Horror        1680 non-null   int64 
 14  Musical       1680 non-null   int64 
 15  Mystery       1680 non-null   int64 
 16  Romance       1680 non-null   int64 
 17  Sci-Fi        1680 non-null   int64 
 18  Thriller      1680 non-null   int64 
 19  War   

title and release date are objects . rest are ints

In [10]:
movie.describe()

movie id       Action    Adventure    Animation    Childrens  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean    841.525595     0.149405     0.080357     0.025000     0.072619   
std     485.609591     0.356593     0.271926     0.156171     0.259587   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%     421.750000     0.000000     0.000000     0.000000     0.000000   
50%     841.500000     0.000000     0.000000     0.000000     0.000000   
75%    1261.250000     0.000000     0.000000     0.000000     0.000000   
max    1682.000000     1.000000     1.000000     1.000000     1.000000   

            Comedy        Crime  Documentary        Drama      Fantasy  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.300595     0.064881     0.029762     0.431548     0.013095   
std       0.458653     0.246389     0.169980     0.495440     0.113717   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Film-Noir       Horror      Musical      Mystery      Romance  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.014286     0.054762     0.033333     0.036310     0.147024   
std       0.118701     0.227583     0.179559     0.187115     0.354235   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi     Thriller          War      Western  
count  1680.000000  1680.000000  1680.000000  1680.000000  
mean      0.060119     0.149405     0.042262     0.016071  
std       0.237778     0.356593     0.201246     0.125788  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000

this is meaningless

**Ratings**

In [11]:
rating.head()

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

We notice 4 cols a user id , movie id , a rating and a timestamp

In [12]:
rating.shape

(100000, 4)

There are 100,000 records and 4 cols

In [13]:
rating.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

The ratings range from 1 to 5 with 4 being the median and an avg of 3.52

In [14]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user id    100000 non-null  int64
 1   movie id   100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


all cols are ints

**How many movies belong to a particular genre?**

In [15]:
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [16]:
rating_list = ['Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western']

In [17]:
movie[rating_list].sum().sort_values(ascending=False)

Drama          725
Comedy         505
Action         251
Thriller       251
Romance        247
Adventure      135
Childrens      122
Crime          109
Sci-Fi         101
Horror          92
War             71
Mystery         61
Musical         56
Documentary     50
Animation       42
Western         27
Film-Noir       24
Fantasy         22
dtype: int64

Here are the genres and the counts sorted in desc order

**Can we find the movies that have more than one genre?**

In [18]:
movie['Number of Genres'] = movie.loc[:,'Action':'Western'].sum(axis=1)
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Film-Noir  Horror  Musical  \
0          1       1      0            0  ...          0       0        0   
1          0       0      0            0  ...          0       0        0   
2          0       0      0            0  ...          0       0        0   
3          0       1      0            0  ...          0       0        0   
4          0       0      1            0  ...          0       0        0   

   Mystery  Romance  Sci-Fi  Thriller  War  Western  Number of Genres  
0        0        0       0         0    0        0                 3  
1        0        0       0         1    0        0                 3  
2        0        0       0         1    0        0                 1  
3        0        0       0         0    0        0                 3  
4        0        0       0         1    0        0                 3  

[5 rows x 22 columns]

In [19]:
movie[movie['Number of Genres']>1][['movie title','Number of Genres']].sort_values(ascending=False,by='Number of Genres').head(5).style.hide()

We add a column that counts the no of genres

**Can we find the genres that are most liked by the users?**

In [20]:
df_merge = movie.merge(rating,on='movie id',how="inner")
df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Mystery  Romance  Sci-Fi  Thriller  War  \
0       1      0            0  ...        0        0       0         0    0   
1       1      0            0  ...        0        0       0         0    0   
2       1      0            0  ...        0        0       0         0    0   
3       1      0            0  ...        0        0       0         0    0   
4       1      0            0  ...        0        0       0         0    0   

   Western  Number of Genres  user id  rating  timestamp  
0        0                 3      308       4  887736532  
1        0                 3      287       5  875334088  
2        0                 3      148       4  877019411  
3        0                 3      280       4  891700426  
4        0                 3       66       3  883601324  

[5 rows x 25 columns]

In [23]:
rating_list

['Action',
 'Adventure',
 'Animation',
 'Childrens',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Fantasy',
 'Film-Noir',
 'Horror',
 'Musical',
 'Mystery',
 'Romance',
 'Sci-Fi',
 'Thriller',
 'War',
 'Western']

In [31]:
for i in rating_list:
    print(i,':' , df_merge[df_merge[i]==1].rating.mean())

Action : 1.0
Adventure : 1.0
Animation : 1.0
Childrens : 1.0
Comedy : 1.0
Crime : 1.0
Documentary : 1.0
Drama : 1.0
Fantasy : 1.0
Film-Noir : 1.0
Horror : 1.0
Musical : 1.0
Mystery : 1.0
Romance : 1.0
Sci-Fi : 1.0
Thriller : 1.0
War : 1.0
Western : 1.0


Can we find which movies are rated most of the times ?

In [36]:
movie_count = df_merge.groupby(['movie title'])['rating'].count().reset_index()
movie_count.sort_values(ascending=False,by='rating').head(5).style.hide()

Star Wars,Contact and Fargo are the top 3 rated most

extract some relationship between demographic details of the users and ratings for the movies?

In [39]:
df_user_rating = user.merge(rating,on='user id',how='inner')
df_user_rating.groupby(['gender'])['rating'].mean()

gender
F    3.531507
M    3.529289
Name: rating, dtype: float64

Split between male and female is pretty much the same for ratings

In [41]:
df_user_rating.groupby(['occupation'])['rating'].mean()

occupation
administrator    3.635646
artist           3.653380
doctor           3.688889
educator         3.670621
engineer         3.541407
entertainment    3.441050
executive        3.349104
healthcare       2.896220
homemaker        3.301003
lawyer           3.735316
librarian        3.560781
marketing        3.485641
none             3.779134
other            3.552377
programmer       3.568260
retired          3.466750
salesman         3.582944
scientist        3.611273
student          3.515143
technician       3.532230
writer           3.375723
Name: rating, dtype: float64

Splits by jobs

In [57]:

df_user_rating_sorted = df_user_rating.groupby(['movie id','zip code'])['rating'].mean().sort_values(ascending=False).to_frame()

df_user_rating_sorted.merge(movie,on='movie id',how='inner')[['movie title','rating']].head(5).style.hide()